File: 967-minimum-falling-path-sum/minimum-falling-path-sum.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        n = len(matrix)
        if n == 1:
            return matrix[0][0]

        for r in range(1, n):
            matrix[r][0] += min(matrix[r - 1][0], matrix[r - 1][1])
            matrix[r][-1] += min(matrix[r - 1][-1], matrix[r - 1][-2])
            for c in range(1, n - 1):
                 matrix[r][c] += min(matrix[r - 1][c - 1],
                                     matrix[r - 1][c],
                                     matrix[r - 1][c + 1])
    
        return min(matrix[-1])

File: 967-minimum-falling-path-sum/minimum-falling-path-sum.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        for (int r = 1; r < n; r++) {
            matrix[r][0] += Math.min(matrix[r - 1][0], matrix[r - 1][1]);
            matrix[r][n - 1] += Math.min(matrix[r - 1][n - 1], matrix[r - 1][n - 2]);
            for (int c = 1; c < n - 1; c++) {
                matrix[r][c] += Math.min(matrix[r - 1][c - 1], Math.min(
                                        matrix[r - 1][c],
                                        matrix[r - 1][c + 1]));
            }
        }
        int minSum = matrix[n - 1][0];
        for (int c = 1; c < n; c++) {
            minSum = Math.min(minSum, matrix[n - 1][c]);
        }
        return minSum;
    }
}